File: legacy/cb/db/repositories/resource_repo.py
from __future__ import annotations
import time
from pathlib import Path
from typing import List

from cb.db.connection import get_connection

def track_resource(
    resource_type: str,
    name: str,
    profile_name: str,
    controller_name: str = "",
    db_path: Path | None = None,
) -> None:
    now = int(time.time())
    conn = get_connection(db_path)
    try:
        conn.execute(
            """INSERT OR REPLACE INTO user_resources 
               (resource_type, name, profile_name, controller_name, created_at)
               VALUES (?, ?, ?, ?, ?)""",
            (resource_type, name, profile_name, controller_name, now)
        )
        conn.commit()
    finally:
        conn.close()

def untrack_resource(
    resource_type: str,
    name: str,
    profile_name: str,
    controller_name: str = "",
    db_path: Path | None = None,
) -> None:
    conn = get_connection(db_path)
    try:
        conn.execute(
            """DELETE FROM user_resources 
               WHERE resource_type = ? AND name = ? AND profile_name = ? AND controller_name = ?""",
            (resource_type, name, profile_name, controller_name)
        )
        conn.commit()
    finally:
        conn.close()

def get_tracked_resources(
    resource_type: str,
    profile_name: str,
    controller_name: str = "",
    db_path: Path | None = None,
) -> List[str]:
    conn = get_connection(db_path)
    try:
        rows = conn.execute(
            """SELECT name FROM user_resources 
               WHERE resource_type = ? AND profile_name = ? AND controller_name = ?""",
            (resource_type, profile_name, controller_name)
        ).fetchall()
        return [r["name"] for r in rows]
    finally:
        conn.close()

File: src/core/db/repositories/resource-repo.ts
/**
 * User-resource tracking repository.
 * Behavioural 1:1 port of legacy/cb/db/repositories/resource_repo.py
 */

import { getConnection } from "../connection";

interface ResourceNameRow {
  name: string;
}

/**
 * Insert or replace a tracked resource record.
 * Mirrors Python track_resource().
 */
export function trackResource(
  resourceType: string,
  name: string,
  profileName: string,
  controllerName = "",
  dbPath?: string
): void {
  const now = Math.floor(Date.now() / 1000);
  const db = getConnection(dbPath);
  try {
    db.run(
      `INSERT OR REPLACE INTO user_resources
         (resource_type, name, profile_name, controller_name, created_at)
         VALUES (?, ?, ?, ?, ?)`,
      [resourceType, name, profileName, controllerName, now]
    );
  } finally {
    db.close();
  }
}

/**
 * Remove a tracked resource record.
 * Mirrors Python untrack_resource().
 */
export function untrackResource(
  resourceType: string,
  name: string,
  profileName: string,
  controllerName = "",
  dbPath?: string
): void {
  const db = getConnection(dbPath);
  try {
    db.run(
      `DELETE FROM user_resources
       WHERE resource_type = ? AND name = ? AND profile_name = ? AND controller_name = ?`,
      [resourceType, name, profileName, controllerName]
    );
  } finally {
    db.close();
  }
}

/**
 * Return names of all tracked resources matching the given type, profile, and
 * optional controller.
 * Mirrors Python get_tracked_resources().
 */
export function getTrackedResources(
  resourceType: string,
  profileName: string,
  controllerName = "",
  dbPath?: string
): string[] {
  const db = getConnection(dbPath);
  try {
    const rows = db.query<ResourceNameRow, [string, string, string]>(
      `SELECT name FROM user_resources
       WHERE resource_type = ? AND profile_name = ? AND controller_name = ?`
    ).all(resourceType, profileName, controllerName);
    return rows.map((r) => r.name);
  } finally {
    db.close();
  }
}
